### PR TITLE
Introduce `update_user_directory_on`

### LINKED ISSUE #10339: Shared rooms API doesn't work if the worker handling the request doesn't update the user directory
#7785 added `^/_matrix/client/unstable/uk.half-shot.msc2666/user/shared_rooms/.*$`.

We check the `update_user_directory` config flag when handling this API, to basically sanity check that the tables we're about to query are roughly up-to-date (`users_who_share_private_rooms`, `users_in_public_rooms`). However, if the request is handled on a *different* synapse worker which does not have that config flag set, then the request is wrongly refused.

I'm not sure if there is an easy way to check if another Synapse process is configured to update the user directory, or whether we should just remove the check. Another option would just be to ensure this endpoint is handled on the same worker.

### PR DESCRIPTION
**Pull request by ShadowJonathan**
Fixes #10339

This changes the `update_user_directory` config variable to a `update_user_directory_on` variable, which more explicitly points or denotes which process is supposed to be working user directory updates. This defaults to the main process.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

`Signed-off-by: Jonathan de Jong <jonathan@automatia.nl>`

### CODE REVIEW

File: synapse/config/server.py
Comment (ShadowJonathan) [inline synapse/config/server.py:326]: Should we keep this config option for now (for backwards compatibility), but let it emit a warning if defined?

What should we do if it has a `False` value? Default to "synapse.app.user_dir"?

**Comment by ShadowJonathan** [comment]
Blocked on https://github.com/matrix-org/sytest/issues/1173 to let sytest issues proceed correctly.

File: synapse/config/server.py
Comment (ShadowJonathan) [reply synapse/config/server.py:326]: I think with a `False` value, we should interpret it as "undefined", unless the local worker is the user_dir worker, then it'll suddenly "match", this should be enough to be backwards-compatible, but we should emit a compatibility warning when this happens.

Comment (clokep) [reply synapse/config/server.py:326]: We would probably want to have some sort of backwards compatibility here, yes. We usually do that when updating the configuration.

Comment (ShadowJonathan) [reply synapse/config/server.py:326]: Alright, but do you think the backwards compatibility method described in my second comment looks alright?

**Comment by ShadowJonathan** [comment]
https://github.com/matrix-org/sytest/pull/1174 unblocks this, decided its easier to just introduce the config change, than process the other issue.

**Comment by ShadowJonathan** [comment]
I added backwards compatibility to this, this is unblocked now.

**Comment by turt2live** [comment]
I have not looked at the patch, but it'd be quite nice to keep the disabling side effect of the flag. In my use case, t2bot.io doesn't update the user directory at all and there is no worker for it, so the original intent of the setting was to save CPU/DB cycles. 

I will admit that it's a use case with sample size 1 though.

**Comment by ShadowJonathan** [comment]
*Technically* it would still work, as no string can match `False`, so consider that a feature :)

**Comment by turt2live** [comment]
can it get documented as a feature please? 😇 

Relying on magic behaviour isn't great.

File: synapse/app/admin_cmd.py
Comment (squahtx) [inline synapse/app/admin_cmd.py:214]:
```suggestion
    config.server.worker_to_update_user_directory = False
```
Should this be `False`?

File: synapse/config/server.py
Comment (squahtx) [inline synapse/config/server.py:40]: Is `UNDEFINED` purely internal to this file?
ie. Will outside code never observe `ServerConfig.worker_to_update_user_directory` with this value?

File: synapse/config/server.py
Comment (squahtx) [inline synapse/config/server.py:334]: Could we add a type annotation to document the expected set of values?
something like `Union[None, False, str, ANY_USER_DIRECTORY_WORKER]`

File: synapse/app/generic_worker.py
Comment (squahtx) [inline synapse/app/generic_worker.py:461]: I think there's a preference for having imports at the top of the file where possible.

File: synapse/rest/client/shared_rooms.py
Comment (ShadowJonathan) [reply synapse/rest/client/shared_rooms.py:57]: Yes, see https://github.com/matrix-org/synapse/issues/10339#issuecomment-965523795 and https://github.com/matrix-org/synapse/issues/10339#issuecomment-971509019.

This makes the admin explicitly select a worker to handle this, and de-magicifies the "oh, the user directory handles this if its false", with this value, any worker (if it is specified and running) can handle the update, the sample config is updated to reflect this behaviour, and warn the admin that it's their responsibility to see that the worker handling it is online.

The default value is `None`, meaning the main process handles this.

File: synapse/config/server.py
Comment (ShadowJonathan) [reply synapse/config/server.py:334]: The last one isn't a type, so that wouldn't pass mypy

File: synapse/app/generic_worker.py
Comment (ShadowJonathan) [reply synapse/app/generic_worker.py:461]: This is to avoid cyclic dependencies, but i think it could be okay enough, let me check.

File: synapse/config/server.py
Comment (ShadowJonathan) [reply synapse/config/server.py:40]: No, though i'll add an `assert` to make that sure

File: synapse/rest/client/shared_rooms.py
Comment (squahtx) [reply synapse/rest/client/shared_rooms.py:57]: I see, thanks for explaining!

File: synapse/config/server.py
Comment (ShadowJonathan) [reply synapse/config/server.py:334]: Yup, i'll add a comment though

image

Comment (squahtx) [reply synapse/config/server.py:334]: That's very true. Elsewhere in the code we appear to create a single entry enum: https://github.com/matrix-org/synapse/blob/develop/synapse/util/caches/cached_call.py#L72

Though if we're going to make an enum we might as well roll the "main process" and "nobody" values into it.

Comment (squahtx) [reply synapse/config/server.py:334]: A comment is a good alternative!

**Review by squahtx — APPROVED** [review]
lgtm!

**Comment by squahtx** [comment]
@ShadowJonathan The user directory sytests without workers seem to be failing after I merged the sytest PR. Could you investigate?

File: docs/sample_config.yaml
Comment (richvdh) [inline docs/sample_config.yaml:154]: please have a look at https://github.com/matrix-org/synapse/blob/develop/docs/code_style.md#configuration-file-format and ensure this matches.

File: docs/sample_config.yaml
Comment (richvdh) [inline docs/sample_config.yaml:153]: so it can be a string, or a boolean, or `null`? ew. can we find a way to limit it to just one type or `null`?

File: docs/sample_config.yaml
Comment (richvdh) [reply docs/sample_config.yaml:154]: It should probably be in the `workers` section?

Comment (richvdh) [reply docs/sample_config.yaml:154]: and maybe it should be called `update_user_directory_on` for consistency with `run_background_tasks_on` ?

File: docs/sample_config.yaml
Comment (richvdh) [inline docs/sample_config.yaml:153]: why would you want *no* process to handle the user directory? why would that be useful?

File: docs/sample_config.yaml
Comment (richvdh) [reply docs/sample_config.yaml:154]: (should it default to the same value as `run_background_tasks_on`?)

**Review by richvdh — COMMENTED** [review]
sorry, just some drive-by comments from looking at the sample config. Afraid I haven't looked at anything else!

**Comment by richvdh** [comment]
> Afraid I haven't looked at anything else!

oh, but others have, so yay.

**Review by richvdh — CHANGES_REQUESTED** [review]
config setting needs updates, as above

File: docs/sample_config.yaml
Comment (ShadowJonathan) [reply docs/sample_config.yaml:153]: We could, the extra type was suggested by @turt2live to basically "disable" the config, though maybe we can change that to be `""`, an empty string.

Comment (ShadowJonathan) [reply docs/sample_config.yaml:153]: See [this comment](https://github.com/matrix-org/synapse/pull/11450#issuecomment-987109870)

File: docs/sample_config.yaml
Comment (ShadowJonathan) [reply docs/sample_config.yaml]: I've addressed all three first issues, but the last issue is still unresolved, i'm leaving this thread open for discussion on that.

File: docs/sample_config.yaml
Comment (richvdh) [inline docs/sample_config.yaml:2664]: this still doesn't match the conventions specified at https://github.com/matrix-org/synapse/blob/develop/docs/code_style.md#configuration-file-format. Please go and read it.

**Review by richvdh — CHANGES_REQUESTED** [review]

File: docs/sample_config.yaml
Comment (richvdh) [reply docs/sample_config.yaml:2664]: I guess it's only fair to give you some clues.

> A line consisting of only `#`.
> ...
> For other options, the example should give some non-default value which is likely to be useful to the reader.

File: synapse/config/workers.py
Comment (richvdh) [inline synapse/config/workers.py:34]: convention is to call things like this `SENTINEL` rather than `UNDEFINED`, though tbh I think it would be better to avoid the need for it with an explicit test like:

```
if 'update_user_directory' in config:
```

File: synapse/config/workers.py
Comment (richvdh) [inline synapse/config/workers.py:38]: can we not avoid the need for this ugliness by deciding whether we should update the user directory in this module whether we should update the user dir (in the same way that we do with `run_background_tasks`, `start_pushers`, `send_federation`, etc)?

It feels like we've invented a completely different wheel when we already have a perfectly functional one.

**Review by richvdh — CHANGES_REQUESTED** [review]

File: synapse/config/workers.py
Comment (ShadowJonathan) [reply synapse/config/workers.py:34]: Ah, wasn't aware of that trick, will apply it. :+1:

File: synapse/config/workers.py
Comment (ShadowJonathan) [reply synapse/config/workers.py:38]: You mean something along the lines of checking for `worker_app` here if `update_user_directory` is defined, and then setting `update_user_directory_on = worker_name` depending on that?

File: docs/sample_config.yaml
Comment (ShadowJonathan) [reply docs/sample_config.yaml:2664]: Ah gotcha, will update. :+1:

File: synapse/config/workers.py
Comment (richvdh) [reply synapse/config/workers.py:38]: no, I mean getting rid of `config.worker.update_user_directory_on` altogether, and just having a boolean `config.worker.update_user_directory`, in much the same way as we have a `config.worker.send_federation`.

Comment (ShadowJonathan) [reply synapse/config/workers.py:38]: That discussion is out of scope for this PR, please comment on https://github.com/matrix-org/synapse/issues/10339 with that, as this PR is applying the suggestions made there.

Comment (richvdh) [reply synapse/config/workers.py:38]: I can't see anything about the implementation in #10339? It seems to just be talking about the format of the config file (which I'm not suggesting you change).

Comment (ShadowJonathan) [reply synapse/config/workers.py:38]: https://github.com/matrix-org/synapse/issues/10339 talks about the disposition that we can't be sure the user directory will be updated at any time, I suggested to make the config value accept None to indicate explicitly that the user directory will be updated on the main process, and any string value for any worker name.

@anoadragon453 approved of this at the time, this PR is executing what is said in that issue, if I were to move ahead on this PR, then I need to know the consensus, so please discuss on that issue if you think this is not a good idea.

Comment (reivilibre) [reply synapse/config/workers.py:38]: @ShadowJonathan I think Rich is not referring to making any config format change, but instead that the config object only stores true/false whether **this** worker will run them.

I refer to `run_background_tasks` (one of the fields Rich mentioned):
```python
        background_tasks_instance = config.get("run_background_tasks_on") or "master"
        self.run_background_tasks = (
            self.worker_name is None and background_tasks_instance == "master"
        ) or self.worker_name == background_tasks_instance
```

Put in other words:
A Synapse process doesn't care exactly *who (which process)* is updating the directory. It only cares whether *it's* the one doing it or not.

Comment (ShadowJonathan) [reply synapse/config/workers.py:38]: That snippet has the same behaviour as this current config change, it doesn't test for true/false, are you sure you got the right snippet?

Comment (ShadowJonathan) [reply synapse/config/workers.py:38]: With some private clarification from rich (after i saw we were miscommunicating on this issue), i'll alter the internal variable to `update_user_directory: bool`, which is derived from the yaml config value `update_user_directory_on: Option[str]`, with backwards compat from the yaml config option `update_user_directory`

File: synapse/app/generic_worker.py
Comment (richvdh) [inline synapse/app/generic_worker.py:465]: errr this seems a bit of an odd constraint. why isn't the user directory worker allowed to do the user directory updates?

(the previous code was checking that user directory updates were disabled on the *main* process before they were enabled implicitly on a worker by setting `worker_app: synapse.app.user_dir`)

Comment (richvdh) [reply synapse/app/generic_worker.py:465]: also note: if you're updating the text of the warning, use a `"""` at the top rather than this clunky multiple-lines-of-`"` thing.

File: synapse/config/workers.py
Comment (richvdh) [inline synapse/config/workers.py:342]: this seems a very verbose, and not very clear, way of writing this logic.

```suggestion
        if "update_user_directory" in config:
            # Backwards compatibility case
            logger.warning(USER_UPDATE_DIRECTORY_DEPRECATION_WARNING)

            if config["update_user_directory"]:
                # Main process should update
                update_user_directory_on = "master"
            else:
                # user directory worker should update
                update_user_directory_on = "synapse.app.user_dir"
        else:
            update_user_directory_on = config.get("update_user_directory_on") or "master"
        
        self.update_user_directory = (
            is_master_process and update_user_directory_on == "master"
        ) or self.worker_name == update_user_directory_on
```

File: tests/utils.py
Comment (richvdh) [inline tests/utils.py:172]: if the intention is to set this to a non-existent worker, might be better to make that clearer:

```suggestion
        "update_user_directory_on": "non-existent worker",
```

File: tests/handlers/test_user_directory.py
Comment (richvdh) [inline tests/handlers/test_user_directory.py:59]: suggest a comment here (and below) to clarify what exactly this is doing.

**Review by richvdh — CHANGES_REQUESTED** [review]
this is definitely much clearer and easier to follow now, thanks! A few notes though.

File: synapse/config/workers.py
Comment (richvdh) [inline synapse/config/workers.py:307]: this should probably be fixed to use `instance_name` while you're here:

```suggestion
            self.instance_name == background_tasks_instance
        )
```

File: synapse/config/workers.py
Comment (ShadowJonathan) [reply synapse/config/workers.py:314]: It's to explicitly set out the intended type, and this'll also snag mypy if any of the branches does not set this variable.

File: synapse/app/generic_worker.py
Comment (ShadowJonathan) [reply synapse/app/generic_worker.py:465]: Good point, should i turn this into a warning then, to log when `update_user_directory_on` isnt specified to the local worker (if it's a user_dir one)?

File: synapse/config/workers.py
Comment (ShadowJonathan) [reply synapse/config/workers.py:342]: This'd make the above `self.update_user_directory: bool` assertion also redundant.

File: synapse/app/generic_worker.py
Comment (richvdh) [inline synapse/app/generic_worker.py:130]: right, I see the confusion here now.

"not allowed to" sounds like I've tried to enable something invalid.  Maybe `s/not allowed to/not configured to/`.

But I don't really understand why this is a thing we need to warn about. Why is it a problem if the updates are done on the main process?

Note that the current wording sounds very much like an error, indicating something that definitely won't work.

File: synapse/app/generic_worker.py
Comment (richvdh) [inline synapse/app/generic_worker.py:467]: given that this happens before logging is set up, I think it probably just ends up in stderr anyway. It may be best to stick to `sys.stderr.write` for consistency.

File: synapse/config/workers.py
Comment (richvdh) [inline synapse/config/workers.py:304]: suggest explicit parens here for clarity
```suggestion
        self.run_background_tasks = (self.instance_name == background_tasks_instance)
```

File: synapse/config/workers.py
Comment (richvdh) [inline synapse/config/workers.py:331]: really quite sure that the `:bool` is redundant here

File: tests/handlers/test_user_directory.py
Comment (richvdh) [inline tests/handlers/test_user_directory.py:59]: this isn't terribly clear.
```suggestion
        # Enable user directory updates
```

File: tests/handlers/test_user_directory.py
Comment (richvdh) [inline tests/handlers/test_user_directory.py:1018]: this still needs a comment

**Review by richvdh — CHANGES_REQUESTED** [review]

File: synapse/config/workers.py
Comment (ShadowJonathan) [reply synapse/config/workers.py:304]: Black clears those, iirc

File: tests/handlers/test_user_directory.py
Comment (ShadowJonathan) [reply tests/handlers/test_user_directory.py:59]: Yes, but this comment exists to explain *why* the user updates flag is set to true there.

File: synapse/app/generic_worker.py
Comment (ShadowJonathan) [reply synapse/app/generic_worker.py:130]: > Maybe s/not allowed to/not configured to/.

👍

> Why is it a problem if the updates are done on the main process?

That's not the problem, the problem is is that the configuration does not appoint the local worker to do the updates.

The config can now appoint any worker to do this job by its name, so the job of the user directory worker isn't that unique anymore, but it's uniqueness is still supported by synapse.

> Note that the current wording sounds very much like an error

That's because it is, the local worker has a "type" of being a designated user update worker, while the configuration does not allow it to do just that. This is more or less a warning that "your config, as-is, is contradictory"

The previous version exited the process with `1` when hitting this point, should that be re-introduced?

File: synapse/config/workers.py
Comment (ShadowJonathan) [reply synapse/config/workers.py:304]: Yeah it does, not much i can do here

**Comment by richvdh** [comment]
Ugh, I remain hopelessly confused about what is going on here, and how it fixes the original issue (https://github.com/matrix-org/synapse/issues/10339). I have commented there.

**Comment by ShadowJonathan** [comment]
I'll be closing this PR in light of the retrospective at https://github.com/matrix-org/synapse/issues/10339#issuecomment-1039175679, where it became clear this PR is the wrong solution to this problem.

**Pull request closed without merging**
